extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -0,0 +1,235 @@
+                           .signature(stageRowSignature)
+                           .maxWorkerCount(maxWorkerCount)
+                           .shuffleSpec(nextShuffleSpec)
+                           .processorFactory(new WindowOperatorQueryFrameProcessorFactory(
+                               queryToRun,
+                               operatorList.get(i),
+                               stageRowSignature
+                           ))
+        );
+      }
+    }
+    return queryDefBuilder.queryId(queryId).build();
+  }
+
+  private boolean validateAndReturnOperatorList(
+      WindowOperatorQuery originalQuery,
+      List<List<OperatorFactory>> operatorList
+  )
+  {
+    final List<OperatorFactory> operators = originalQuery.getOperators();
+    List<OperatorFactory> operatorFactoryList = new ArrayList<>();
+    for (OperatorFactory of : operators) {
+      operatorFactoryList.add(of);
+      if (of instanceof WindowOperatorFactory) {
+        operatorList.add(operatorFactoryList);
+        operatorFactoryList = new ArrayList<>();
+      } else if (of instanceof NaivePartitioningOperatorFactory) {
+        if (((NaivePartitioningOperatorFactory) of).getPartitionColumns().isEmpty()) {
+          operatorList.clear();
+          operatorList.add(originalQuery.getOperators());
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private ShuffleSpec findShuffleSpecForNextWindow(List<OperatorFactory> operatorFactories, int maxWorkerCount)
+  {
+    NaivePartitioningOperatorFactory partition = null;
+    NaiveSortOperatorFactory sort = null;
+    List<KeyColumn> keyColsOfWindow = new ArrayList<>();
+    for (OperatorFactory of : operatorFactories) {
+      if (of instanceof NaivePartitioningOperatorFactory) {
+        partition = (NaivePartitioningOperatorFactory) of;
+      } else if (of instanceof NaiveSortOperatorFactory) {
+        sort = (NaiveSortOperatorFactory) of;
+      }
+    }
+    Map<String, ColumnWithDirection.Direction> colMap = new HashMap<>();
+    if (sort != null) {
+      for (ColumnWithDirection sortColumn : sort.getSortColumns()) {
+        colMap.put(sortColumn.getColumn(), sortColumn.getDirection());
+      }
+    }
+    assert partition != null;
+    if (partition.getPartitionColumns().isEmpty()) {
+      return null;
+    }
+    for (String partitionColumn : partition.getPartitionColumns()) {
+      KeyColumn kc;
+      if (colMap.containsKey(partitionColumn)) {
+        if (colMap.get(partitionColumn) == ColumnWithDirection.Direction.ASC) {
+          kc = new KeyColumn(partitionColumn, KeyOrder.ASCENDING);
+        } else {
+          kc = new KeyColumn(partitionColumn, KeyOrder.DESCENDING);
+        }
+      } else {
+        kc = new KeyColumn(partitionColumn, KeyOrder.ASCENDING);
+      }
+      keyColsOfWindow.add(kc);
+    }
+    return new HashShuffleSpec(new ClusterBy(keyColsOfWindow, 0), maxWorkerCount);
```

Comment:
Why are we doing hashShuffling ? Shoudn't it be sort merge shuffle spec. 
since we would require keys to be sorted in a window  for RAC calculation

Reply:
I see that's why you are adding the keyOrder while generating the clusterByCol. 
@LakshSingla We can also add partition boosting but that can be done as a follow up.

Reply:
Yes, we can add that as a followup